### PR TITLE
feat(optimistic-oracle): integrate ancillary data into optimistic oracle

### LINKED ISSUE #2238: Add ancillary data to optimistic oracle
**Why**

To support ancillary data wherever users access the DVM.

**How**

Just by adding a passthrough arg to its interfaces.

**Resolution**

All OO functions have an ancillary data argument that is ultimately passed through to the DVM.

**Difficulty Score [1-10]**

- 5: ~4 hours

### PR DESCRIPTION
**Pull request by mrice32**
**Motivation**

Optimistic oracle should support ancillary data as the DVM does.

**Summary**

- All optimistic oracle methods now take ancillary data.
- Optimistic oracle now "stamps" all requests it sends to the DVM with ancillary data: "OptimisticOracle[REQUESTING_CONTRACT_ADDRESS]". This will allow the DVM to choose to respond invalid to any contracts that request through the optimistic oracle and aren't "playing by the rules".

**Issue(s)**

Fixes #2238

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/FeePayer.sol [hunk 1]
```diff
@@ -178,7 +178,8 @@ abstract contract FeePayer is AdministrateeInterface, Testable, Lockable {
             FixedPoint.Unsigned memory collateralPool = _pfc();
 
             // The final fee must be < available collateral or the fee will be larger than 100%.
-            require(collateralPool.isGreaterThan(amount), "Final fee is more than PfC");
+            // Note: revert reason removed to save bytecode.
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/FeePayer.sol:181]: :(

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol [hunk 2]
```diff
@@ -304,4 +321,12 @@ abstract contract FundingRateApplier is FeePayer {
         // cumulative funding rate multiplier.
         newCumulativeFundingRateMultiplier = currentCumulativeFundingRateMultiplier.mul(unsignedPeriodMultiplier);
     }
+
+    function _getAncillaryData() public view returns (bytes memory) {
```
Comment (chrismaree) [inline packages/core/contracts/financial-templates/common/FundingRateApplier.sol:325]: nit: if this is a public method it should probs be with the other funding methods and have natspec/reentrancy guards?

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol [hunk 3]
```diff
@@ -523,4 +588,9 @@ contract OptimisticOracle is Testable, Lockable {
     function _getIdentifierWhitelist() internal view returns (IdentifierWhitelistInterface) {
         return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
     }
+
+    // Stamps the ancillary data blob with the optimistic oracle tag denoting what contract requested it.
+    function _stampAncillaryData(bytes memory ancillaryData, address requester) internal pure returns (bytes memory) {
+        return abi.encodePacked(ancillaryData, "OptimisticOracle", requester);
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:594]: this seems good for now. we should propose an UMIP at some point soon that defines the expected structure of ancillary Data to standardize it.

File: packages/core/contracts/oracle/implementation/OptimisticOracle.sol [hunk 4]
```diff
@@ -7,7 +7,7 @@ import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
 import "@openzeppelin/contracts/math/SafeMath.sol";
 
 import "../interfaces/StoreInterface.sol";
-import "../interfaces/OracleInterface.sol";
+import "../interfaces/OracleAncillaryInterface.sol";
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/OptimisticOracle.sol:10]: really nice that this is the only import change needed :)

**Review by chrismaree — APPROVED** [review]
This is awesome! It's not too complex a change and fortunately, it's almost as simple as adding one additional param through the optimistic oracle. Tests look good and overall I'm happy with this change.

File: packages/core/contracts/financial-templates/common/FundingRateApplier.sol (hunk 2 above)
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:325]: it's not. Making it public was accidental!

Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol:325]: Changing it now :)

Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FundingRateApplier.sol]: Done

File: packages/core/contracts/financial-templates/common/FeePayer.sol (hunk 1 above)
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/FeePayer.sol:181]: yeah :/

**Review by daywiss — APPROVED** [review]
Looks good, nice job

**Pull request merged**